### PR TITLE
Add files to describe benchmarks in python

### PR DESCRIPTION
**Pull request by pzread**
Files to describe the benchmarks in python.

The python scripts to define benchmarks will go under `build_tools/benchmarks/configs`

### CODE REVIEW

**Comment by GMNGeoffrey** [comment]
I think the distinction has been that `build_tools/benchmarks` is all the infra related to the benchmarks and `benchmarks/` is for the actual definition of the benchmarks. Kind of like how tests are defined with the source code and the scripts to run them are under `build_tools`. Not sure what the right split is here, but I think we should only have both directories if we can justify them both and why they both exist. Otherwise I'd be inclined to put everything under `build_tools/benchmarks`

**Comment by pzread** [comment]
> I think the distinction has been that `build_tools/benchmarks` is all the infra related to the benchmarks and `benchmarks/` is for the actual definition of the benchmarks. Kind of like how tests are defined with the source code and the scripts to run them are under `build_tools`. Not sure what the right split is here, but I think we should only have both directories if we can justify them both and why they both exist. Otherwise I'd be inclined to put everything under `build_tools/benchmarks`

Thinking for a while, it's probably okay to put all executable scripts at the first level of `build_tools/benchmarks` and define the benchmarks under `build_tools/benchmarks/configs`. All scripts refer to `build_tools/benchmarks` as the root python module path (since the executable scripts are at the first level, they will have this module path by default)

**Comment by GMNGeoffrey** [comment]
From our offline conversation, I thought the goal was to be able to share Python utilities between various scripts and stuff (e.g. I think we've written a half-dozen "get the git root" and "execute this command but maybe print it first" utility functions). I'm fine saying that isn't worth it though

**Comment by pzread** [comment]
After the discussion, I was playing around some ways to make the executable scripts including the utilities directory in the search path. It works but not very elegant. Maybe there are other ways.

I didn't find an example sharing utilities between benchmarks and other build scripts in IREE. If there is or will be such a case, I'm happy to explore the idea of a single utilities directory more : ) Otherwise I think it isn't reasonable enough to make it more complicated.

It's true that having the utilities directory in the search path also helps the structure of the benchmark directory, as we don't need to put all executable scripts at the first level. We can have `benchmarks/tools` and `benchmarks/workflow_scripts`, etc. for executable scripts with different purposes. I don't feel too bad to put all of them at the first level, but if you think this is a good improvement, we can try the utilities directory and the search path stuff.

**Comment by GMNGeoffrey** [comment]
Ah, I see you're avoiding subdirectories entirely. :shrug: We don't have any examples of sharing utilities because we don't have a way to share them :stuck_out_tongue: I have rewritten the same utility a few times though

**Comment by pzread** [comment]
Sent out #10065 and see how other people think about it 🙂

**Comment by pzread** [comment]
I think we can separate this PR from #10065. We put the code under `build_tools/benchmarks` for now and move to the new place once #10065 is resolved.

File: build_tools/benchmarks/definitions/common.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+class DeviceSpec(object):
+  """Benchmark device specification."""
+  id: str
+  # Device vendor name. E.g., Pixel-6.
+  vendor_name: str
+  architecture: DeviceArchitecture
+  platform: DevicePlatform
+  # Device-specific configurations. E.g., big-cores. Benchmark machines use
+  # these configs to set up the devices.
+  configs: List[str]
```
Comment (GMNGeoffrey) [inline build_tools/benchmarks/definitions/common.py:81]: The other things are all properties of the device. This is a property of how we run something on the device. Do we want to have them together like this?

File: build_tools/benchmarks/definitions/iree.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from dataclasses import dataclass
+from enum import Enum
+from typing import List
+
+from definitions.common import DeviceArchitecture, DevicePlatform, DeviceSpec, Model, ModelInputData
```
Comment (GMNGeoffrey) [inline build_tools/benchmarks/definitions/iree.py:11]: Couldn't this just be "from common"?

**Review by GMNGeoffrey — APPROVED** [review]

File: build_tools/benchmarks/definitions/iree.py (hunk 2 above)
Comment (pzread) [reply build_tools/benchmarks/definitions/iree.py]: Changed to relative import `from .common`. `from common` doesn't work if the module is imported from other directory.

File: build_tools/benchmarks/definitions/common.py (hunk 1 above)
Comment (pzread) [reply build_tools/benchmarks/definitions/common.py]: I have an idea about including these "software configurable" device properties in the device spec. On a heterogeneous processor, little cores and big cores can be seen as the different devices. This field allows us to define device specs like `Pixel-6_4-big-cores`, `Pixel-6_4-little-cores`, ... So we can use the same 4-threads run config to benchmark on different 4-cores specs.

Also when comparing with other ML frameworks, even they have different run configs to specify 4-threads parallelism, they can refer to the same device spec, so we know they are run on the same big/little cores setting.

**Pull request merged**